crypto/threads_pthread.c

```diff
@@ -42,12 +45,562 @@
+         * updates prior to the load.  This is a non-issue on cache coherent
+         * systems like x86, but is relevant on other arches
+         * Note: This applies to the reload below as well
+         */
+        qp_idx = (uint64_t)ATOMIC_LOAD_N(&lock->reader_idx, __ATOMIC_ACQUIRE);
+
+        /*
+         * Notes of use of __ATOMIC_RELEASE
+         * This counter is only read by the write side of the lock, and so we
+         * specify __ATOMIC_RELEASE here to ensure that the write side of the
+         * lock see this during the spin loop read of users, as it waits for the
+         * reader count to approach zero
+         */
+        ATOMIC_ADD_FETCH(&lock->qp_group[qp_idx].users, VAL_READER,
+                         __ATOMIC_RELEASE);
+
+        /* if the idx hasn't changed, we're good, else try again */
+        if (qp_idx == (uint64_t)ATOMIC_LOAD_N(&lock->reader_idx, __ATOMIC_ACQUIRE))
+            break;
+
+        /*
+         * Notes on use of __ATOMIC_RELEASE
+         * As with the add above, we want to ensure that this decrement is
+         * seen by the write side of the lock as soon as it happens to prevent
+         * undue spinning waiting for write side completion
+         */
+        ATOMIC_SUB_FETCH(&lock->qp_group[qp_idx].users, VAL_READER,
+                         __ATOMIC_RELEASE);
+    }
+
+    return &lock->qp_group[qp_idx];
+}
+
+void ossl_rcu_read_lock(CRYPTO_RCU_LOCK lock)
+{
+    struct rcu_thr_data *data;
+    int i, available_qp = -1;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    if (data == NULL) {
+        data = OPENSSL_zalloc(sizeof(*data));
+        OPENSSL_assert(data != NULL);
+        CRYPTO_THREAD_set_local(&rcu_thr_key, data);
+        ossl_init_thread_start(NULL, NULL, free_rcu_thr_data);
+    }
+
+    for (i = 0; i < MAX_QPS; i++) {
+        if (data->thread_qps[i].qp == NULL && available_qp == -1)
+            available_qp = i;
+        /* If we have a hold on this lock already, we're good */
+        if (data->thread_qps[i].lock == lock)
+            return;
+    }
+
+    /*
+     * if we get here, then we don't have a hold on this lock yet
+     */
+    assert(available_qp != -1);
+
+    data->thread_qps[available_qp].qp = get_hold_current_qp(lock);
+    data->thread_qps[available_qp].lock = lock;
+}
+
+void ossl_rcu_read_unlock(CRYPTO_RCU_LOCK lock)
+{
+    int i;
+    struct rcu_thr_data *data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+
+    assert(data != NULL);
+
+    for (i = 0; i < MAX_QPS; i++) {
+        if (data->thread_qps[i].lock == lock) {
+            /*
+             * As with read side acquisition, we use __ATOMIC_RELEASE here
+             * to ensure that the decrement is published immediately 
+             * to any write side waiters
+             */
+            ATOMIC_SUB_FETCH(&data->thread_qps[i].qp->users, VAL_READER,
```

Comment:
Theres no guarantee of what the value in users will be here.  If there are other readers holding the lock, it will be > 0 and thats an ok/expected condition.  Note data in this function is thread local storage, but the qp it points to is potentially shared by multiple threads.

Reply:
Surely, before we subtract 1 from it users the value must be > 0. Otherwise we will get an underflow. Something must have gone wrong in that case. I'm just suggesting adding an assert to check for that case.

Reply:
sorry. the way I read your initial comment, I thought you were suggesting that we check for a zero condition after the sub, which made no sense to me.  Yes, we could do an underflow check here

Reply:
added